apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/PgNumeric.cs

```diff
@@ -0,0 +1,487 @@
+        private static readonly Regex s_validation = new Regex(@"^-?[0-9]+\.?[0-9]*$", RegexOptions.Compiled);
+
+        // Note: the following properties must be declared *after* s_maxValue and s_minValue. Initialization order matters.
+
+        /// <summary>
+        /// Zero represented as a <see cref="PgNumeric"/>. This is the default value for the type.
+        /// </summary>
+        public static PgNumeric Zero { get; } = default;
+
+        /// <summary>
+        /// NaN (Not a number) represented as a <see cref="PgNumeric"/>. 
+        /// </summary>
+        public static PgNumeric NaN { get; } = Parse(NotANumber);
+
+        /// <summary>
+        /// The maximum valid value for <see cref="PgNumeric"/>, equal to ((BigInteger.Pow(10, MaxPrecision) - 1) / (BigInteger.Pow(10, MaxScale)))
+        /// </summary>
+        public static PgNumeric MaxValue { get; } = new PgNumeric(s_maxValue);
+
+        /// <summary>
+        /// The minimum valid value for <see cref="PgNumeric"/>, equal to -((BigInteger.Pow(10, MaxPrecision) - 1) / (BigInteger.Pow(10, MaxScale)))
+        /// </summary>
+        public static PgNumeric MinValue { get; } = new PgNumeric(s_minValue);
+
+        // The scaled value of PgNumeric would be stored in this variable.
+        private readonly BigInteger _integer;
+
+        // This flag would be true if PgNumeric is not a number.
+        private readonly bool _notANumber;
+
+        // The scale of numeric value stored.
+        private readonly int _scale;
+
+        private PgNumeric(BigInteger integer, int scale = 0, bool isNotANumber = false)
+        {
+            if (!IsRawValueInRange(integer))
+            {
+                throw new OverflowException("PgNumeric value out of range");
+            }
+
+            _scale = scale;
+            _notANumber = isNotANumber;
+            _integer = integer;
+        }
+
+        private static bool IsRawValueInRange(BigInteger integer) => integer >= s_minValue && integer <= s_maxValue;
+
+        /// <summary>
+        /// Compares this value with <paramref name="other"/>.
+        /// </summary>
+        /// <param name="other">The value to compare with this one</param>
+        /// <returns>A negative integer if this value is less than <paramref name="other"/>; a positive integer
+        /// if it's greater than <paramref name="other"/>; zero if they're equal.</returns>
+        public int CompareTo(PgNumeric other)
+        {
+            // Check for NaN
+            if (_notANumber)
+            {
+                return other._notANumber ? 0 : 1;
+            }
+
+            // if scale is same, we can compare the integer value
+            if (_scale == other._scale)
+            {
+                return _integer.CompareTo(other._integer);
+            }
+
+            // return comparision of normalized value.
+            return Normalize(this).NormalizedValue.CompareTo(Normalize(other).NormalizedValue);
```

Comment:
Actually we might need to "normalize" to MaxScale always, so that we have a consistent way of calculating the HashCode, else Equals and HashCode won't match.

We chatted yesterday and we said that maybe storing the numbers in MaxScale always might hit performance as we would, for the most part, carry a lot of zeros, around order of 10^4. I still think that's a bad idea. But if for HashCode we need to normalize to MaxScale then not storing the number fully scaled in the first place solves very little.

An alternative is to represent the number slightly different:
- BigInteger always without trailing zeros
- Scale that accepts negative values

Then we don't carry extra zeros around. HashCode is trivial, i.e. the combination of the parts, and comparison can still be done by temporarily scaling to the biggest scale of the two operands as my code above. @jskeet what do you think?